src/csharp/SDK/Image.cs

```diff
@@ -1,349 +1,343 @@
                     }
+
+                    return (void*)this._Buffer;
                 }
             }
         }
-
-        private IntPtr _Buffer = IntPtr.Zero;
-
         
-        public unsafe void* Buffer
+        /// <summary>
+        /// Gets the Memory containing the image data.
+        /// </summary>
+        public unsafe Memory<byte> Memory
         {
             get
             {
-                if (_Buffer != IntPtr.Zero)
+                lock (this)
                 {
+
                     if (disposedValue)
                         throw new ObjectDisposedException(nameof(Image));
 
-                    return (void*)_Buffer;
-                }
+                    // If we previously copied the native memory to a managed array, return that array's memory
+                    if (managedBufferCache != null)
+                    {
+                        return new Memory<byte>(managedBufferCache, 0, (int)this.Size);
+                    }
 
-                lock (this)
-                {
-                    if (disposedValue)
-                        throw new ObjectDisposedException(nameof(Image));
+                    // If we previously wrapped the native memory in a IMemoryOwner<T>, return that memory object
+                    if (nativeBufferWrapper != null)
+                    {
+                        return nativeBufferWrapper.Memory;
+                    }
+
+                    IntPtr bufferAddress = (IntPtr)this.Buffer;
 
-                    _Buffer = NativeMethods.k4a_image_get_buffer(handle);
-                    if (_Buffer == IntPtr.Zero)
+                    // If the native buffer is within a memory block that the managed allocator provided,
+                    // return that memory.
+                    Memory<byte> memory = this.Allocator.GetManagedAllocatedMemory(bufferAddress, this.Size);
+                    if (!memory.IsEmpty)
                     {
-                        throw new AzureKinectException("Image has NULL buffer");
+                        return memory;
+                    }
+
+                    // The underlying buffer is not in managed memory.
+                    // We can use one of two strategies to return a Memory<T>
+                    
+                    // If we use the CopyNativeBuffers method, we will allocate and then copy the native memory
+                    // to a managed array, ensuring memory-safe access to the contents of memory, at the expense of
+                    // a memcpy each time we transition the buffer from native to managed, or from managed to native.
+
+                    // If we don't copy the native buffers, we can construct a MemoryManager<T> that wraps that native
+                    // buffer. This has no memcpy cost, but exposes the possibilty of use after free bugs to consumers
+                    // of the library. This is therefore not enabled by default.
+                    if (this.Allocator.SafeCopyNativeBuffers)
```

Comment:
There is some inconsistencies between using this.Allocator and Allocator.Singleton.